cirq/google/serializable_device.py

```diff
@@ -0,0 +1,245 @@
+            gate_set: A SerializableGateSet that can translate the gate_ids
+                into cirq Gates.
+        """
+
+        # Store target sets, since they are refered to by name later
+        allowed_targets: Dict[str, Set[Tuple['cirq.Qid', ...]]] = {}
+        permutation_ids: Set[str] = set()
+        for ts in proto.valid_targets:
+            allowed_targets[ts.name] = cls._create_target_set(ts)
+            if ts.target_ordering == v2.device_pb2.TargetSet.SUBSET_PERMUTATION:
+                permutation_ids.add(ts.name)
+
+        # Store gate definitions from proto
+        gate_defs: Dict[str, v2.device_pb2.GateDefinition] = {}
+        for gs in proto.valid_gate_sets:
+            for gate_def in gs.valid_gates:
+                gate_defs[gate_def.id] = gate_def
+
+        # Loop through serializers and create dictionaries with keys as
+        # the python type of the cirq Gate and duration/targets from the proto
+        durations: Dict[Type['cirq.Gate'], Duration] = {}
+        target_sets: Dict[Type['cirq.Gate'], Set[
+            Tuple['cirq.Qid', ...]]] = dict()
+        permutation_gates: Set[Type['cirq.Gate']] = set()
+        gate_definitions: Dict[Type['cirq.Gate'], _GateDefinition] = {}
+        for gate_type in gate_set.supported_gate_types():
+            for serializer in gate_set.serializers[gate_type]:
+                gate_id = serializer.serialized_gate_id
+                if gate_id not in gate_defs:
+                    raise ValueError(f'Serializer has {gate_id} which is not '
+                                     'supported by the device specification')
+                gate_ts = gate_defs[gate_id].valid_targets
+                which_are_permutations = [t in permutation_ids for t in gate_ts]
+                if any(which_are_permutations):
+                    if not all(which_are_permutations):
+                        msg = f'Id {gate_id} in {gate_set.gate_set_name} ' \
+                            ' mixes SUBSET_PERMUTATION with other types which' \
+                            ' is not currently allowed.'
+                        raise NotImplementedError(msg)
+                    permutation_gates.add(gate_type)
+                gate_picos = gate_defs[gate_id].gate_duration_picos
+                durations[gate_type] = Duration(picos=gate_picos)
+                if gate_type not in target_sets:
+                    target_sets[gate_type] = set()
+                for target_set_name in gate_defs[gate_id].valid_targets:
+                    target_sets[gate_type] |= allowed_targets[target_set_name]
+        for gate_type in target_sets:
+            gate_definitions[gate_type] = _GateDefinition(
+                duration=durations[gate_type],
+                target_set=target_sets[gate_type],
+                is_permutation=(permutation_gates is not None and
+                                gate_type in permutation_gates),
+                number_of_qubits=gate_defs[gate_id].number_of_qubits)
```

Comment:
It looks like you should be able to construct all the `_GateDefinition` objects just from the protos, and put them in a map indexed by gate_id. Then loop over serializers and create a new map indexed by the cirq gate type. This should avoid the need for the `durations`, `target_sets`, and `permutation_gates` temporaries.

Reply:
Not sure how much clearer it is (maybe slightly simpler), but it's done.